pkg/reco/reco.go

```diff
@@ -0,0 +1,237 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"time"
+)
+
+type Recommender interface {
+	Recommend(workloadSpec v1alpha1.WorkloadSpec) (*v1alpha1.HPAConfiguration, error)
+}
+
+type CpuUtilizationBasedRecommender struct {
+	k8sClient    client.Client
+	redLineUtil  float64
+	metricWindow time.Duration
+	scraper      metrics.Scraper
+	metricStep   time.Duration
+	minTarget    int
+	maxTarget    int
+	logger       logr.Logger
+}
+
+func NewCpuUtilizationBasedRecommender(k8sClient client.Client,
+	redLineUtil float64,
+	metricWindow time.Duration,
+	scraper metrics.Scraper,
+	metricStep time.Duration,
+	minTarget int,
+	maxTarget int,
+	logger logr.Logger) *CpuUtilizationBasedRecommender {
+	return &CpuUtilizationBasedRecommender{
+		k8sClient:    k8sClient,
+		redLineUtil:  redLineUtil,
+		metricWindow: metricWindow,
+		scraper:      scraper,
+		metricStep:   metricStep,
+		minTarget:    minTarget,
+		maxTarget:    maxTarget,
+		logger:       logger,
+	}
+}
+
+func (c *CpuUtilizationBasedRecommender) Recommend(workloadSpec v1alpha1.WorkloadSpec) (*v1alpha1.HPAConfiguration,
+	error) {
+
+	end := time.Now()
+	start := end.Add(c.metricWindow)
+
+	dataPoints, err := c.scraper.GetAverageCPUUtilizationByWorkload(workloadSpec.Namespace,
+		workloadSpec.Name,
+		start,
+		end,
+		c.metricStep)
+	if err != nil {
+		c.logger.Error(err, "Error while scraping GetCPUUtilizationBreachDataPoints.")
+		return nil, nil
+	}
+
+	acl, err := c.scraper.GetACLByWorkload(workloadSpec.Namespace, workloadSpec.Name)
+	if err != nil {
+		c.logger.Error(err, "Error while getting GetACL.")
+		return nil, nil
+	}
+
+	perPodResources, err := c.getContainerCPULimitsSum(workloadSpec.Namespace, workloadSpec.Kind, workloadSpec.Name)
+	if err != nil {
+		c.logger.Error(err, "Error while getting getContainerCPULimitsSum")
+		return nil, err
+	}
+
+	optimalTargetUtil, minReplicas, maxReplicas, err := c.findOptimalTargetUtilization(dataPoints,
+		acl,
+		c.minTarget,
+		c.maxTarget,
+		perPodResources)
+	if err != nil {
+		c.logger.Error(err, "Error while executing findOptimalTargetUtilization")
+		return nil, err
+	}
+
+	return &v1alpha1.HPAConfiguration{Min: minReplicas, Max: maxReplicas, TargetMetricValue: optimalTargetUtil}, nil
+}
+
+type TimerEvent struct {
+	Timestamp time.Time
+	Delta     float64
+}
+
+func (c *CpuUtilizationBasedRecommender) simulateHPA(dataPoints []metrics.DataPoint,
+	acl time.Duration,
+	targetUtilization int,
+	perPodResources float64) ([]metrics.DataPoint, int, int, error) {
+
+	if len(dataPoints) == 0 {
+		return []metrics.DataPoint{}, 0, 0, nil
+	}
+	if targetUtilization < 1 || targetUtilization > 100 {
+		return []metrics.DataPoint{}, 0, 0, errors.New(fmt.Sprintf("Invalid value of target utilization: %v."+
+			" Value should be between 1 and 100", targetUtilization))
+	}
+
+	simulatedDataPoints := make([]metrics.DataPoint, len(dataPoints))
+
+	currentReplicas := math.Ceil((dataPoints[0].Value * 100) / float64(targetUtilization) / perPodResources)
```

Comment:
we are assuming here the perPodResources has been constant and is same in the past. Changes to perPodResources in the past is going to mess this up.

Reply:
This is intentional. We are only getting the total cpu utilisation demand curve from the past. Per pod resources don't matter. We only need to get the current perPodResources.